In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# Data

In [5]:
df = pd.read_csv("Speed Dating Data.csv", encoding="ISO-8859-1")

# Liking Scores: The Strategy

$\ell_{ij} = w_1(attractive_{ij} + 1) + w_2(sincere_{ij} + 1) + w_3(intelligent_{ij} + 1) + w_4(fun_{ij} + 1) + w_5(ambitious_{ij} + 1)$

# Find Weights

In [77]:
importances = df[["iid","attr1_1", "sinc1_1", "intel1_1", "fun1_1", "amb1_1"]].groupby("iid").mean()

In [79]:
# scaling importances to be between 0 and 10
importances  = (importances-0)/(10-0)

In [84]:
importances.head()

attr1_1  sinc1_1  intel1_1  fun1_1  amb1_1
iid                                            
1        1.5      2.0       2.0     1.5     1.5
2        4.5      0.5       2.5     2.0     0.0
3        3.5      1.0       3.5     1.0     1.0
4        2.0      2.0       2.0     2.0     1.0
5        2.0      0.5       2.5     2.5     1.0

In [93]:
# deriving weights as the average of how important each person viewed this factor
w_attr = np.round(importances["attr1_1"].mean(),2)
w_sinc = np.round(importances["sinc1_1"].mean(),2)
w_intel = np.round(importances["intel1_1"].mean(),2)
w_fun = np.round(importances["fun1_1"].mean(),2)
w_amb = np.round(importances["amb1_1"].mean(),2)

In [95]:
print("w_attr:  " + str(w_attr))
print("w_sinc:  " + str(w_sinc))
print("w_intel: " + str(w_intel))
print("w_fun:   " + str(w_fun))
print("w_amb:   " + str(w_amb))

w_attr:  2.27
w_sinc:  1.73
w_intel: 2.02
w_fun:   1.75
w_amb:   1.08


# Find Factor Values

We're going to do this by finding how each person rates themself on these factors

In [120]:
values = df[["iid","attr3_1", "sinc3_1", "intel3_1", "fun3_1", "amb3_1"]].groupby("iid").mean()
values.head()

attr3_1  sinc3_1  intel3_1  fun3_1  amb3_1
iid                                            
1        6.0      8.0       8.0     8.0     7.0
2        7.0      5.0       8.0    10.0     3.0
3        8.0      9.0       9.0     8.0     8.0
4        7.0      8.0       7.0     9.0     8.0
5        6.0      3.0      10.0     6.0     8.0

In [135]:
# normalizing these values to have standard normal distribution
values = (values - values.mean()) / values.std()
values.head()

attr3_1   sinc3_1  intel3_1    fun3_1    amb3_1
iid                                                  
1   -0.785746 -0.202528 -0.353867  0.193145 -0.323290
2   -0.066364 -2.327117 -0.353867  1.485544 -2.562565
3    0.653019  0.505668  0.563817  0.193145  0.236529
4   -0.066364 -0.202528 -1.271551  0.839344  0.236529
5   -0.785746 -3.743510  1.481501 -1.099255  0.236529

In [141]:
# impute missing scores using mean
values = values.fillna(0);

In [153]:
# bring iid in as its own column
values["iid"] = values.index
values = values.rename(columns={"attr3_1":"attr",
                                "sinc3_1":"sinc",
                                "intel3_1":"intel",
                                "fun3_1":"fun",
                                "amb3_1":"amb"})
values.head()

attr      sinc     intel       fun       amb  iid
iid                                                       
1   -0.785746 -0.202528 -0.353867  0.193145 -0.323290    1
2   -0.066364 -2.327117 -0.353867  1.485544 -2.562565    2
3    0.653019  0.505668  0.563817  0.193145  0.236529    3
4   -0.066364 -0.202528 -1.271551  0.839344  0.236529    4
5   -0.785746 -3.743510  1.481501 -1.099255  0.236529    5

In [156]:
# export to CSV
values.to_csv("people_scores.csv")

# Final Formulation for Liking Scores:

$\ell_{ij}$ = 2.27( people_scores.loc[j,"attr"] + 1) + 1.73( people_scores.loc[j,"sinc"] + 1) + 2.02( people_scores.loc[j,"intel"] + 1) + 1.75( people_scores.loc[j,"fun"] + 1) + 1.08( people_scores.loc[j,"amb"] + 1)